modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C

```diff
@@ -0,0 +1,337 @@
+  auto input_mesh = dynamic_cast<ReplicatedMesh *>(_mesh.get());
+  _input_mesh_external_bid =
+      MooseMeshUtils::getBoundaryID(_input_mesh_external_boundary, *input_mesh);
+
+  // Use CoM of the input mesh as its origin for azimuthal calculation
+  const Point origin_pt = MooseMeshUtils::meshCentroidCalculator(*input_mesh);
+  const Real origin_x = origin_pt(0);
+  const Real origin_y = origin_pt(1);
+  // Vessel for containing maximum radius of the boundary nodes
+  Real max_input_mesh_node_radius;
+
+  if (!isBoundaryValid(
+          *input_mesh, max_input_mesh_node_radius, origin_pt, _input_mesh_external_bid))
+    paramError("input_mesh_external_boundary",
+               "This mesh generator does not work for the provided boundary.");
+  if (max_input_mesh_node_radius >= _peripheral_ring_radius)
+    paramError(
+        "peripheral_ring_radius",
+        "The peripheral ring to be generated must be large enough to cover the entire input mesh.");
+  if (!isExternalBoundary(*input_mesh, _input_mesh_external_bid))
+    paramError("input_mesh_external_boundary",
+               "The boundary provided is not an external boundary.");
+
+  std::vector<Point> input_ext_bd_pts;
+  std::vector<Point> output_ext_bd_pts;
+  std::vector<std::tuple<Real, Point, Point>> azi_points;
+  std::vector<Real> azi_array;
+  Real tmp_azi(0.0);
+  auto mesh = buildReplicatedMesh(2);
+
+  std::vector<std::tuple<dof_id_type, unsigned short int, boundary_id_type>> side_list =
+      input_mesh->get_boundary_info().build_side_list();
+  input_mesh->get_boundary_info().build_node_list_from_side_list();
+  std::vector<std::tuple<dof_id_type, boundary_id_type>> node_list =
+      input_mesh->get_boundary_info().build_node_list();
+
+  // 2D vector containing all the node positions of the peripheral region
+  std::vector<std::vector<Point>> points_array(_peripheral_layer_num + 1,
+                                               std::vector<Point>(node_list.size()));
+  // 2D vector containing all the node ids of the peripheral region
+  std::vector<std::vector<dof_id_type>> node_id_array(_peripheral_layer_num + 1,
+                                                      std::vector<dof_id_type>(node_list.size()));
+  // Node counter for the external boundary
+  unsigned int input_ext_node_num = 0;
+
+  // Loop over all the boundary nodes
+  for (unsigned int i = 0; i < node_list.size(); ++i)
+    if (std::get<1>(node_list[i]) == _input_mesh_external_bid)
+    {
+      input_ext_node_num++;
+      // Define nodes on the inner and outer boundaries of the peripheral region.
+      input_ext_bd_pts.push_back(input_mesh->point(std::get<0>(node_list[i])));
+      tmp_azi = atan2(input_ext_bd_pts.back()(1) - origin_y, input_ext_bd_pts.back()(0) - origin_x);
+      output_ext_bd_pts.push_back(Point(_peripheral_ring_radius * std::cos(tmp_azi),
+                                        _peripheral_ring_radius * std::sin(tmp_azi),
+                                        0.0));

```

Comment:
0 might not be the z-coordinate of the input mesh

Reply:
Will allow nonzero z coordinate meshes.